checklist/src/main/scala/checklist/Checked.scala

```diff
@@ -0,0 +1,138 @@
+    fold(
+      (es, w) => Errored(es.map(f), w),
+      Warned(_, _),
+      Succeeded(_)
+    )
+
+  def flatMap[S](f: R => Checked[E, W, S]): Checked[E, W, S] =
+    fold(
+      Errored(_, _),
+      (_, result) => combineWith(f(result))((_, s) => s),
+      result => combineWith(f(result))((_, s) => s)
+    )
+
+  def value: Option[R] = fold((_, _) => None, (_, r) => Some(r), Some(_))
+  def warnings: List[W] = fold((_, w) => w, (w, _) => w.toList, _ => Nil)
+  def errors: List[E] = fold((errs, _) => errs.toList, (_, _) => Nil, _ => Nil)
+
+  def iorValidated: List[W] Ior Validated[NonEmptyList[E], R] =
+    fold(
+      (es, ws) => Ior.both(ws, Validated.invalid(es)),
+      (ws, result) => Ior.both(ws.toList, Validated.valid(result)),
+      result => Ior.right(Validated.valid(result))
+    )
+
+  def destructured: (List[E], List[W], Option[R]) =
+    fold(
+      (errs, warns) => (errs.toList, warns, None),
+      (warns, result) => (Nil, warns.toList, Some(result)),
+      result => (Nil, Nil, Some(result))
+    )
+
+  def isValid: Boolean = fold((_, _) => false, (_, _) => false, _ => true)
+  def hasAValue: Boolean = fold((_, _) => false, (_, _) => true, _ => true)
+  def hasErrors: Boolean = fold((_, _) => true, (_, _) => false, _ => false)
+  def hasNoErrors: Boolean = fold((_, _) => false, (_, _) => true, _ => true)
+  def hasWarnings: Boolean = fold((_, w) => w.nonEmpty, (_, _) => true, _ => false)
+  def hasNoWarnings: Boolean = fold((_, w) => w.isEmpty, (_, _) => false, _ => true)
+}
+
+object Checked {
+  def succeeded[E, W, R](result: R): Checked[E, W, R] = Succeeded(result)
+
+  def errored[E, W, R](errors: NonEmptyList[E]): Checked[E, W, R] = Errored(errors, Nil)
+  def erroredWithWarnings[E, W, R](errors: NonEmptyList[E], warnings: List[W]): Checked[E, W, R] = Errored(errors, warnings)
+
+  def warned[E, W, R](warnings: NonEmptyList[W], result: R): Checked[E, W, R] = Warned(warnings, result)
+
+  implicit def checkedMonad[E, W]: Monad[Checked[E, W, ?]] =
+    new Monad[Checked[E, W, ?]] {
+      def ap[A, B](ff: Checked[E, W, A => B])(fa: Checked[E, W, A]): Checked[E, W, B] = ff.combineWith(fa)(_ apply _)
+      def pure[A](a: A): Checked[E, W, A] = Succeeded(a)
+      def map[A, B](fa: Checked[E, W, A])(f: A => B): Checked[E, W, B] = fa.map(f)
+      def flatMap[A, B](fa: Checked[E, W, A])(f: A => Checked[E, W, B]): Checked[E, W, B] = fa.flatMap(f)
+    }
+
+  implicit def checkedMonoid[E, W, A](implicit monoid: Monoid[A]): Monoid[Checked[E, W, A]] =
+    new Monoid[Checked[E, W, A]] {
+      def empty: Checked[E, W, A] = Succeeded(monoid.empty)
+      def combine(l: Checked[E, W, A], r: Checked[E, W, A]): Checked[E, W, A] = l.combine(r)(monoid)
+    }
+
+}
+
+case class Errored[E, W, R](errors: NonEmptyList[E], warnings: List[W]) extends Checked[E, W, R]
+case class Warned[E, W, R](warnings: NonEmptyList[W], result: R) extends Checked[E, W, R]
+case class Succeeded[E, W, R](result: R) extends Checked[E, W, R]
```

Comment:
Is hard coding Nels and Lists above really useful, or is it overconstraining?  I like the benefits it brings within the context of this lib, but they may not be necessary.

Reply:
I guess people may want to implement their own list-like type to replace `NonEmptyList`? For example:

```scala
trait MyErrors
trait MyWarnings

type MyChecked[A] = Checked[MyErrors, MyWarnings, A]
```

Reply:
I had a quick play around with my own noddy implementation (didn't get anywhere near what @Jacoby6000 has accomplished though) and it quickly became obvious that leaving `E` and `W` types completely free is *very* useful - as long as I can provide `Semigroup` for accumulation, I should be able to provide whatever error / warning type I like.